compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEUnrollAndDistributeWorkgroup.cpp

```diff
@@ -0,0 +1,507 @@
+  IRRewriter rewriter(moduleOp.getContext());
+  WalkResult res = moduleOp->walk([&](AMDAIE::WorkgroupOp workgroupOp) {
+    WalkResult workgroupRes = workgroupOp->walk([&](scf::ForallOp forallOp) {
+      if (failed(scf::forallToForLoop(rewriter, forallOp))) {
+        workgroupOp.emitOpError()
+            << "failed to transform scf.forall to scf.for";
+        return WalkResult::interrupt();
+      }
+      return WalkResult::advance();
+    });
+    if (workgroupRes.wasInterrupted()) return WalkResult::interrupt();
+    return WalkResult::advance();
+  });
+  if (res.wasInterrupted()) return failure();
+  return success();
+}
+
+/// Hoist an affine apply op on a scf.for op's induction variable into that
+/// scf.for block.
+LogicalResult hoistAffineApplyDependingOnFor(ModuleOp moduleOp) {
+  IRRewriter rewriter(moduleOp.getContext());
+  moduleOp->walk([&](affine::AffineApplyOp applyOp) {
+    (void)hoistForAffineApplyOp(rewriter, applyOp);
+  });
+  return success();
+}
+
+/// Unroll the scf.for loops selected for parallel execution on the AIE
+/// array. Try to hoist dma ops that don't depend on the loops' induction
+/// variables to avoid duplicated copies.
+///
+/// This rewriter consists of a sequence of transformations:
+///   1) First, try to promote the for loop if possible.
+///   2) Try hoisting dma ops outside the scf.for operation.
+///   3) Unroll and distribute the logical objectfifos remaining in the scf.for
+///   loop.
+class AMDAIEUnrollWorkgroupLoops : public OpRewritePattern<scf::ForOp> {
+  using OpRewritePattern<scf::ForOp>::OpRewritePattern;
+
+  /// Hoist dma ops outside the scf.for operation if there are no dependencies
+  /// on the scf.for loop, or on other dmas producing or consuming data from the
+  /// one considered.
+  ///
+  /// NOTE: This method makes an assumption that the operations
+  /// happen on the most local memory and therefore dmas moving data into more
+  /// local memory can be hoisted before the scf.for loop. And on the other
+  /// hand, dmas moving data away from local memory can be hoisted behind the
+  /// scf.for. This assumption could be removed in the future.
+  template <typename Iterator>
+  LogicalResult hoistDmaOps(PatternRewriter &rewriter, scf::ForOp forOp) const {
+    // Keep track of whether a hoist happened.
+    bool hoistHappened{false};
+
+    // Create utility function to check whether an operand depends on a scf.for
+    // induction variable or a value within the scf.for's scope
+    auto dependsOnLoop = [&](OpOperand &operand) -> bool {
+      Operation *op = operand.get().getDefiningOp();
+      if (!op) return operand.get() == forOp.getInductionVar();
+
+      // Check for an induction var and whether the parent scf.for is the same
+      // as the one we're using for the hoist
+      auto parentForOp = op->getParentOfType<scf::ForOp>();
+      return (operand.get() == forOp.getInductionVar()) ||
+             (parentForOp && parentForOp == forOp);
```

Comment:
Wouldn't just `parentForOp == forOp` work ?

Reply:
Yes, updated it